lib/middleware/operation.js

```diff
@@ -32,55 +36,92 @@ function findPropertyOperations ({ types, method, term, resource }) {
   return operations
 }
 
-function factory (api) {
-  return async (req, res, next) => {
-    if (!req.hydra.resource) {
-      return next()
+function mapOperations ({ api, res, method }) {
+  return res.locals.hydra.resources.flatMap((resource) => {
+    let moreOperations
+    if ('property' in resource) {
+      moreOperations = findPropertyOperations({ resource, method, api }).toArray()
+    } else {
+      const types = clownface({ ...api, term: [...resource.types] })
+      moreOperations = findClassOperations(types, method).toArray()
+    }
+
+    if (moreOperations.length) {
+      return moreOperations.map(operation => ({ resource, operation }))
     }
 
+    return [{ resource, operation: null }]
+  })
+}
+
+function findOperation (operationMiddleware) {
+  return async function (req, res, next) {
     const method = req.method === 'HEAD' ? 'GET' : req.method
-    const types = clownface({ ...api, term: [...req.hydra.resource.types] })
 
-    let operations
-    if (req.hydra.term.equals(req.hydra.resource.term)) {
-      // only look for direct operation when the root resource is requested
-      operations = findClassOperations(types, method)
+    let operations = mapOperations({ api: req.hydra.api, res, method })
+      .filter(({ operation }) => operation)
+
+    if (operationMiddleware) {
+      operations = operationMiddleware(operations, { req, res })
+    }
```

Comment:
Calling the operations middleware hook every time. This way an API might get a chance to select an operation even if none were found otherwise. Or replace those found with something else completely.